fem/gslib.hpp

```diff
@@ -290,6 +293,60 @@ class OversetFindPointsGSLIB : public FindPointsGSLIB
    using FindPointsGSLIB::Interpolate;
 };
 
+/** \brief  Class for gather-scatter (gs) operations on Vectors based on
+    corresponding global identifiers.
+    This functionality is useful for gs-ops on
+    DOF values across processor boundary, where the global identifier would be
+    the corresponding true DOF index. Operations currently supported are
+    min, max, sum, and multiplication. Note: identifier 0 does not participate
+    in the gather-scatter operation.
+    For example, consider a vector, v:
+    - v = [0.3, 0.4, 0.25] on rank1,
+    - v = [0.6, 0.1] on rank 2,
+    - v = [-0.2, 0.3, 0.7, 0.] on rank 3.
+    Consider a corresponding Array<int>, a:
+    - a = [1, 2, 3] on rank 1,
+    - a = [3, 2] on rank 2,
+    - a = [1, 2, 0, 3] on rank 3.
+    A gather-scatter "minimum" operation, done as follows:
+    GSOPGSLIB gs = GSOPGSLIB(MPI_COMM_WORLD, a);
+    gs.GS(v, GSOp::MIN);
+    would return into v:
+    - v = [-0.2, 0.1, 0.] on rank 1,
+    - v = [0., 0.1] on rank 2,
+    - v = [-0.2, 0.1, 0.7, 0] on rank 3,
+    where the values have been compared across all processors based on the
+    integer identifier. */
+class GSOPGSLIB
+{
+protected:
+   struct gslib::crystal *cr;               // gslib's internal data
+   struct gslib::comm *gsl_comm;            // gslib's internal data
+   struct gslib::gs_data *gsl_data = NULL;
+   int num_ids;
+
+public:
+   GSOPGSLIB(Array<long long> &ids);
+
+#ifdef MFEM_USE_MPI
+   GSOPGSLIB(MPI_Comm comm_, Array<long long> &ids);
+#endif
+
+   virtual ~GSOPGSLIB();
+
+   /// Supported operation types. See class description.
+   enum GSOp {ADD, MUL, MIN, MAX};
+
+   /// Update the identifiers used for the gather-scatter operator.
+   /// Same @a ids get grouped together and id == 0 does not participate.
+   /// See class description.
+   virtual void UpdateIdentifiers(Array<long long> &ids);
```

Comment:
Do you need these to be `virtual`? I'm getting warnings from the calls to this in the constructors (those calls will always use the current-class version).

Reply:
This does not need to be virtual.